### ISSUE TITLE
System.ArgumentOutOfRangeException

### ISSUE THREAD

**Issue by pluma9** [PowerShell/PSReadLine#2272]
### Environment
PSReadLine: 2.2.0-beta1
PowerShell: 7.1.2
OS: Microsoft Windows 10.0.18363
BufferWidth: 82
BufferHeight: 52

Last 200 Keys
```
 r i Tab Enter c o p y - w o Tab Enter UpArrow Escape UpArrow UpArrow UpArrow DownArrow Enter s c r Tab Spacebar l i n Tab Enter UpArrow Enter Escape d d i UpArrow Enter UpArrow Enter UpArrow Enter UpArrow Enter UpArrow Enter s c r i Tab Enter s c r i Tab Spacebar l i Tab Enter UpArrow Enter UpArrow Enter UpArrow Enter Escape Escape k Escape d d " h e l Escape A h e l l o Spacebar w o r l d " Spacebar - c o n Tab Spacebar " l l " Enter Escape k w w w w w d e i l i k Tab Escape b d e i c o n Tab Escape w w x Enter Escape k w w w w w d e i l i Tab Spacebar Backspace Escape l w Enter Escape k w w f l I Escape f l f l f Escape A Escape h i * Escape l l i * Enter Escape / l l i Backspace Backspace i s t Backspace Backspace s t Enter A Ctrl+Backspace Ctrl+Backspace Ctrl+Backspace Ctrl+Backspace Escape Escape d d
```

### Exception
```
System.ArgumentOutOfRangeException: The value must be greater than or equal to zero and less than the console's buffer size in that dimension. (Parameter 'top')
Actual value was 52.
   at System.ConsolePal.SetCursorPosition(Int32 left, Int32 top)
   at Microsoft.PowerShell.PSConsoleReadLine.ReallyRender(RenderData renderData, String defaultColor)
   at Microsoft.PowerShell.PSConsoleReadLine.ForceRender()
   at Microsoft.PowerShell.PSConsoleReadLine.DeleteLine(Nullable`1 key, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.ProcessOneKey(PSKeyInfo key, Dictionary`2 dispatchTable, Boolean ignoreIfNoAction, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.ViChordHandler(Dictionary`2 secondKeyDispatchTable, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.ViChord(Nullable`1 key, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.ProcessOneKey(PSKeyInfo key, Dictionary`2 dispatchTable, Boolean ignoreIfNoAction, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.InputLoop()
   at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Runspace runspace, EngineIntrinsics engineIntrinsics, CancellationToken cancellationToken)
```

**Comment by daxian-dbw** [comment]
@pluma9 Thanks for reporting this issue. Did you see this in VS Code terminal? Can you please provide repro steps?

**Comment by bruttium** [comment]
I'm not the original poster, but I see something *similar* but not identical when I use "ctrl-space" for auto completion.  I mention this here in case it fosters ideas, but I'm not saying it's exactly the same thing.  Note, I get this sort of error if I type something like "Get-" followed by Ctrl+Space.  Or even  "Get-A" followed by Ctrl+Space.
This is in Windows Terminal, with Windows PowerShell 5.1.  It doesn't happen if I just open PowerShell on its own.

```
-----------------------------------------------------------------------
Oops, something went wrong.  Please report this bug with the details below.
Report on GitHub: https://github.com/lzybkr/PSReadLine/issues/new
-----------------------------------------------------------------------
Last 6 Keys:
 g e t - Ctrl+Space y

Exception:
System.ArgumentOutOfRangeException: The value must be greater than or equal to zero and less than the console's buffer size in that dimension.
Parameter name: top
Actual value was -337.
   at System.Console.SetCursorPosition(Int32 left, Int32 top)
   at Microsoft.PowerShell.PSConsoleReadLine.Menu.EnsureMenuAndInputIsVisible(IConsole console, Int32 tooltipLineCount)
   at Microsoft.PowerShell.PSConsoleReadLine.Menu.DrawMenu(Menu previousMenu)
   at Microsoft.PowerShell.PSConsoleReadLine.PossibleCompletionsImpl(CommandCompletion completions, Boolean menuSelect)
   at Microsoft.PowerShell.PSConsoleReadLine.CompleteImpl(Boolean menuSelect)
   at Microsoft.PowerShell.PSConsoleReadLine.ProcessOneKey(ConsoleKeyInfo key, Dictionary`2 dispatchTable, Boolean ignoreIfNoAction, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.InputLoop()
   at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Runspace runspace, EngineIntrinsics engineIntrinsics)
-----------------------------------------------------------------------
```

**Comment by iSazonov** [comment]
I can repo (Get-<ctrl-Space>) in Windows Terminal with _Windows PowerShell 5.1_ and PSReadline **2.0.0 Beta2**.
I see the same PSReadline 2.2.0 Beta2.
But all works well with _PowerShell 7.1.2_.

**Comment by pluma9** [comment]
> @pluma9 Thanks for reporting this issue. Did you see this in VS Code terminal? Can you please provide repro steps?

No, I see this in Windows Terminal (version 1.7.572.0). I'm not sure if this helps, but here's my PSReadLine configuration:

```powershell
Set-PSReadLineOption -BellStyle None
Set-PSReadLineOption -EditMode Vi
Set-PSreadLineOption -PredictionSource HistoryAndPlugin
Set-PSreadLineOption -PredictionViewStyle InlineView
Set-PSReadLineKeyHandler  -Function SwitchPredictionView -Chord 'F2'
```

For repro steps, I haven't tried typing those last 200 keys. I encountered this while playing around with a few Powershell commands.

**Comment by daxian-dbw** [comment]
@bruttium The issue you ran into was already fixed (the exception was the same but it's from a different call stack). Please upgrade to the [2.1.0 version of PSReadLine](https://www.powershellgallery.com/packages/PSReadLine/2.1.0) from PowerShell Gallery. See the [upgrading section](https://github.com/PowerShell/PSReadLine#upgrading) for instructions. Please let us know if you run into the same issue with the latest version.

@iSazonov The 2.0.0-beta2 version was very old, and most of the `ArgumentOutOfRangeException` you may run into with that version should have been fixed. If you run into this exception at the same call stack like reported in this issue, it's something we want to fix but couldn't because we cannot reproduce the issue. If you have a repro, that would be great and please share the steps.

**Comment by bruttium** [comment]
@daxian-dbw Thanks for the followup.  Yes, I'm not a master of keeping all of the versions of PowerShell straight and up to date.  Updating PSReadLine did correct the problem. I appreciate your follow-up, and apologize if I was just ignorant. Cheers.

> @bruttium The issue you ran into was already fixed (the exception was the same but it's from a different call stack). Please upgrade to the [2.1.0 version of PSReadLine](https://www.powershellgallery.com/packages/PSReadLine/2.1.0) from PowerShell Gallery. See the [upgrading section](https://github.com/PowerShell/PSReadLine#upgrading) for instructions. Please let us know if you run into the same issue with the latest version.

**Comment by iSazonov** [comment]
Sorry it is another issue.
**It is for 2.2.0-Beta2, Windows Terminal and Windows PowerShell.** (Works well in standalone Windows PowerShell) It seems it was reported as fixed.
Repro steps:
```powershell
Get-Module PSReadline | Remove-Module
ipmo C:\Users\1\Documents\PowerShell\Modules\PSReadLine\2.2.0\PSReadLine
Get-<Ctrl-Space>
```
![image](https://user-images.githubusercontent.com/22290914/111414675-32638780-8702-11eb-8ea9-fe8017e5502a.png)

**Comment by iSazonov** [comment]
As for the issue, I see very strange tail in last 200 keys that is "`e[`e[dd" how could it be? 
"`e[d" sets cursor on top line without clear screen. I don’t understand how this sequence gets inside another one that’s the same.

**Comment by daxian-dbw** [comment]
@bruttium and @iSazonov Thanks for your clarification. I will hide the related conversation make the comment history for the original issue clean.

@iSazonov Do you mean the `Escape d d` keystrokes? It indicates the user was using the VI edit mode.

**Comment by iSazonov** [comment]
> @iSazonov Do you mean the `Escape d d` keystrokes? It indicates the user was using the VI edit mode.

Yes. I don't understand why this comes from the user side. He switched to VI mode with hot keys and got an exception?

I see so many situations where SetCursorPosition raises an exception the reason for which is impossible to understand, that this suggests that we need to come up with some kind of _smart replay_ so that the problem can always be reproduced.

**Comment by endo64** [comment]
I got same error on Windows Terminal Version: 1.7.1033.0

* Hibernate the laptop for a day.
* Turn on the PC , switch desktop and press CTRL-L to clear the screen.

```
Last 116 Keys:
 c d Space t u r Tab t u r k Tab Tab Enter
 c l s Enter
 p h p Space Backspace Backspace Backspace Backspace Backspace Backspace Backspace Backspace C : Ctrl+Alt+\ U s e r s Ctrl+Alt+\ e n d o Ctrl+Alt+\ D o c u m e n t s Ctrl+Alt+\ T u r k t e l e k o m Ctrl+Alt+\ F a k e C D R S e e d e r . r e d Home r e d Space - - c l i Space Enter
 p h p Space a r t i s a n Space s e r v e Enter
 UpArrow Enter
 Ctrl+l

Exception:
System.ArgumentOutOfRangeException: The value must be greater than or equal to zero and less than the console's buffer size in that dimension.
Parameter name: top
Actual value was -1.
   at System.Console.SetCursorPosition(Int32 left, Int32 top)
   at Microsoft.PowerShell.PSConsoleReadLine.ProcessOneKey(ConsoleKeyInfo key, Dictionary`2 dispatchTable, Boolean ignoreIfNoAction, Object arg)
   at Microsoft.PowerShell.PSConsoleReadLine.InputLoop()
   at Microsoft.PowerShell.PSConsoleReadLine.ReadLine(Runspace runspace, EngineIntrinsics engineIntrinsics)
```

**Comment by daxian-dbw** [comment]
@endo64 The `Ctrl+l` issue has been fixed, see #1315 for more details.

**Comment by daxian-dbw** [comment]
Given that we cannot reproduce the issue, there is no further action can be taken at this point.
I will update the issue report to ask for the screenshot when the reported issue happens, which needs include the content on the terminal right before the exception happens. I think that would help the diagnosis of such issues.

For now, I will close this issue. @pluma9 if you run into a similar issue again, please report again, with the screenshot and describe the operations that lead to the exception state, as detailed as possible. Thank you!
